lib/router.js

```diff
@@ -46,10 +46,22 @@ export default class Router {
 
     this.historyMethod = historyMethod;
     this.origChangeStateMethod = this.window.history[this.historyMethod];
-    this.window.history[this.historyMethod] = (...args) => {
-      this.origChangeStateMethod.apply(this.window.history, args);
-      this.navigateToHash();
-    };
+
+    // If this is the true *original* native method, fire pushstate/replacestate events so
+    // that other components can be notified of history changes. Subsequent panel routers
+    // will re-wrap the method but not fire duplicate pushstate/replacestate events.
+    if (this.origChangeStateMethod.name === this.historyMethod) {
```

Comment:
I'm trying to grok how `this.origChangeStateMethod.name === this.historyMethod)` prevents duplicates. is that because the wrapping method has no name, but the original method has a name?

Reply:
exactly. open to any alternate means of detecting this, this is the cleanest I could figure out

Reply:
to be clear, the original has `name === "pushState"` or `name === "replaceState"`; the wrapped functions don't have this because they are declared as anonymous `(...args) => ...`

Reply:
The way we do it in fetch-cache is something like this.

```
  if (window.__origFetch) {
    // already installed
    return;
  }

  window.__origFetch = window.fetch;
window.fetch = async function fetch(url, options) {
```

We store the origin version as a `__method` name and compare the function. Comparing on name is fine, but feels a bit brittle. I'm okay with it for time being tho. We at-least should write some comments about line 53 for future readers.

@tdumitrescu may have opinions.

Reply:
Could we do that here without polluting `window`? Each instance of router only knows of its own `this.origStateChangeMethod` which could be either the native or wrapped version going in; there's no way to compare that to something afaict (unless you store something on window/global, feels more dirty to me). Not sure why you feel this is brittle? [`Function.name`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name) is standard across browsers (sans IE as noted earlier)

I can def elaborate on the comments, I thought the 3 lines there would be sufficient to explain the "why" with the code explaining the "how"

Reply:
@nojvek added 1e3b861583f4da97f0a105c1b4c2a91b13e18f05 to elucidate

Reply:
Thanks